patches/server/0936-Add-Entity-hidden-by-default-flag.patch

```diff
@@ -0,0 +1,207 @@
+discrepancy when trying to getHiddenPlayers - players that are hidden by default are currently 
+not returned from this method, only those hidden with hideEntity or hidePlayer.
+
+Hiding entities by default also respects precedent of keeping entities hidden when other plugins 
+explicitly hide them from a player. If an entity is hidden by default and by a plugin, the entity 
+must have the plugin that hid them from a player show them in order to be shown to a player again. 
+If no plugin had previously hidden the entity before the entity was hidden by default, then any 
+plugin that attempts to show the entity to a player will succeed in doing so.
+
+diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
+index 77c89376495d90d0e7cbf6cd02c9a1c8d9a4340b..c29c2a57b8bab3dd36af0e4ff21c6f0735c260a9 100644
+--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -2277,6 +2277,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+         private final int range;
+         SectionPos lastSectionPos;
+         public final Set<ServerPlayerConnection> seenBy = new ReferenceOpenHashSet<>(); // Paper - optimise map impl
++        public final Set<ServerPlayerConnection> showToEvenWhenHiddenByDefault = new ReferenceOpenHashSet<>(); // Paper - Entity hiddenByDefault
+ 
+         public TrackedEntity(Entity entity, int i, int j, boolean flag) {
+             this.serverEntity = new ServerEntity(ChunkMap.this.level, entity, j, flag, this::broadcast, this.seenBy); // CraftBukkit
+@@ -2385,6 +2386,13 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+                     flag = false;
+                 }
+                 // CraftBukkit end
++
++                // Paper start - Entity hiddenByDefault
++                if (this.entity.isHiddenByDefault() && !this.showToEvenWhenHiddenByDefault.contains(player.connection)) {
++                    flag = false;
++                }
++                // Paper end
++
+                 if (flag) {
+                     if (this.seenBy.add(player.connection)) {
+                         this.serverEntity.addPairing(player);
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index f925a8d550ecbf2044a37bfe58b30d6578c5f6af..dfda6e2f38be417f19dc9d7847f3cae80fb8fac2 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -398,6 +398,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
+     @javax.annotation.Nullable
+     private UUID originWorld;
+     public boolean freezeLocked = false; // Paper - Freeze Tick Lock API
++    public boolean hiddenByDefault; // Paper - Entity hiddenByDefault
```

Comment:
Why is this public if we have setters and getters ?
One of the two should be enough, usually its a public field mutated through the craft bukkit layer directly just for a smaller NMS diff.

Reply:
I agree. Think the thought process here was looking at the surrounding variables. Will change to private.